src/cdsetool/download.py

```diff
@@ -10,33 +10,64 @@
 import tempfile
 import time
 import shutil
+
+from dataclasses import dataclass
+
 from cdsetool._processing import _concurrent_process
 from cdsetool.credentials import Credentials
 from cdsetool.logger import NoopLogger
 from cdsetool.monitor import NoopMonitor
 
 
+@dataclass
+class DownloadResult:
+    """
+    Describes the result of a download operation, whether it was successful or not.
+    Contains relevant information about the feature and the download operation.
+    """
+
+    success: bool
+    feature: dict
+    filename: str
+    message: str
```

Comment:
How about:
```python
feature: dict
filename: Optional[str]
message: Optional[str]

@staticmethod
def ok(feature, filename):
    return DownloadResult(feature, filename, None)

@staticmethod
def fail(feature, message):
    return DownloadResult(feature, None, message)
```
with an invariant that either filename or message is None? Object construction is then done with:
```python
return DownloadResult.ok(feature, filename)
```

I'm not sure about having just `dict`, aren't all keys strings so the type can be `dict[str, Any]`?

Reply:
I like the static contructor methods, but are you sure about using `filename` and `message` as indicators for  success?
Having the extra bool field is much easier to understand, for basically no extra code.

Reply:
[491ebec](https://github.com/SDFIdk/CDSETool/pull/101/commits/491ebecb23f33236af7e03e6b5445b2e4f7d2543)

Reply:
I have only used mypy, but I assume all the type checkers will also require people to write code like this:
```python
rv = download_feature(feature, options)
if not rv.success or rv.filename is None:
   return None
unzip(rv.filename)
```
since the type checker can't figure out the invariant that `success = True <=> filename != None`.

I don't know Python, but is this the use-case for property, so:
```python
@property
def success(self):
     return filename is not None
```
should be used?

Reply:
Assuming the python type checkers are as smart as the TypeScript ones, they should be able to figure out the invariants.
I will configure one or two and have a go at it before i move forward with this

Reply:
I think the property solution would bring the best of both worlds, no derived data stored in the object but it's still possible to do the status check with `rv.success`.

(I've been thinking about suggesting type annotations because I think it would be a good idea, but .. too many other things that also needs to be fixed)